File: 278-first-bad-version/278-first-bad-version.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int first = 0;
        int last = n - 1;
        int mid = 0;
            
        while(first <= last){
            mid = first + (last-first)/2;
            
            if(isBadVersion(mid) == true) {
                if(isBadVersion(mid-1) == false) return mid;
                last = mid-1;
            } 
            else{
                if(isBadVersion(mid+1) == true) return mid + 1;
                first = mid+1;
            }        
        }
        
        return mid;
    } 
    
}

File: 278-first-bad-version/278-first-bad-version.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:
    def firstBadVersion(self, n: int) -> int:
        left = 1
        right = n
        
        while left <= right:
            mid = (left+right)//2
            if isBadVersion(mid) == True:
                if isBadVersion(mid-1) == False:
                    return mid
                else: right = mid-1
            else: left = mid+1
        
                
        